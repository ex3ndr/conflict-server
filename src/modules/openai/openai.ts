import OpenAI from 'openai';

export const openai = new OpenAI({
    apiKey: process.env.OPENAI_TOKEN
});

export async function executeOpenAI(messages: { role: 'user' | 'assistant' | 'system', content: string }[]) {
    let res = await openai.chat.completions.create({
        model: 'gpt-3.5-turbo',
        messages: messages
    });
    return res.choices[0].message.content!;
}