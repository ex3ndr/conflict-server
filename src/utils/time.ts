import { events } from "../app/events";

export async function delay(ms: number) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

export function exponentialBackoffDelay(currentFailureCount: number, minDelay: number, maxDelay: number, maxFailureCount: number) {
    let maxDelayRet = minDelay + ((maxDelay - minDelay) / maxFailureCount) * Math.max(currentFailureCount, maxFailureCount);
    return Math.round(Math.random() * maxDelayRet);
}

export type BackoffFunc = <T>(callback: () => Promise<T>) => Promise<T>;

export function createBackoff(
    opts?: {
        onError?: (e: any, failuresCount: number) => void,
        minDelay?: number,
        maxDelay?: number,
        maxFailureCount?: number
    }): BackoffFunc {
    return async <T>(callback: () => Promise<T>): Promise<T> => {
        let currentFailureCount = 0;
        const minDelay = opts && opts.minDelay !== undefined ? opts.minDelay : 250;
        const maxDelay = opts && opts.maxDelay !== undefined ? opts.maxDelay : 1000;
        const maxFailureCount = opts && opts.maxFailureCount !== undefined ? opts.maxFailureCount : 50;
        while (true) {
            try {
                return await callback();
            } catch (e) {
                if (currentFailureCount < maxFailureCount) {
                    currentFailureCount++;
                }
                if (opts && opts.onError) {
                    opts.onError(e, currentFailureCount);
                }
                let waitForRequest = exponentialBackoffDelay(currentFailureCount, minDelay, maxDelay, maxFailureCount);
                await delay(waitForRequest);
            }
        }
    };
}

export let backoff = createBackoff({ onError: (e) => { console.warn(e); } });

export function delayBreakable(ms: number) {
    // We can cancel delay from outer code
    let promiseResolver: ((value?: any | PromiseLike<any>) => void) | null = null;
    let resolver = () => {
        if (promiseResolver) {
            promiseResolver();
        }
    };
    let promise = new Promise(resolve => {
        promiseResolver = resolve;
        setTimeout(resolve, ms);
    });
    return { promise, resolver };
}

export function resumeKey(key: string) {
    events.emit('pause-' + key);
}

export async function pauseWithKey(ms: number, key: string) {
    let { promise, resolver } = delayBreakable(ms);
    events.addListener('pause-' + key, resolver);
    try {
        await promise;
    } finally {
        events.removeListener('pause-' + key, resolver);
    }
}