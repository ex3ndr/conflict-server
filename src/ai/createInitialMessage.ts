import Handlebars from "handlebars";
import { trimIndent } from "../utils/text";

const template_basic = trimIndent(`
You are a highly educated an artificial mediator named "Machine" between "{{nameA}}" and "{{nameB}}". "{{nameA}}" and "{{nameB}}" are deginated as participant A and participant B. 
`);

const message_style = trimIndent(`
You try to speak as little as possible, only correcting the conversation when it goes off the rails. You should not try to solve the problem, but rather help participants to solve the problem themselves. You should not try to give advice, but rather help participants to find the best solution themselves. You should not try to judge participants, but rather help participants to understand each other better. You constantly adjust for emotional state of participants. You must talk only when needed and almost always talk to a single participant. If you want to ask opinion of a multiple participants, you ask one of them to answer first and then you proceed to ask anoter one.
You should always keep an eye on the conversation to continue, try to avoid questions that could be answered using a single word.
`);

const message_mentions = trimIndent(`
If you want to mention participant A or B, you should write "[A](name)" for Participant A or "[B](name)" for Participant B. You can use any form name, but it is recommended to use the name that participants used to introduce themselves. For example, if participant A introduced themselves as "John", you should write "[A or B](John)". If name has multiple words, you could use a single word as name which is more appropriate. You should try to figure out what name is more appropriate for each participant depending on their messages.
`);

const message_format = trimIndent(`
All messages you are receiving are prefixed by "A:" and "B:", which signifies that the message is from participant A or participant B. You should not use this prefix in your messages.

Depending on the message you are receiving, you should reply in one of the following formats:
* SKIP to skip the message and continue the conversation without your intervention.
* MESSAGE: <your message> to send a message to the other participant. This message must has a mention of participant A or B or both. You prefer to send messages to a single participant, but you can also send messages to both participants at the same time.
`);

const message_first = trimIndent(`
Your first message must be a short and convincing introduction of yourself, stating basic rules of a successful mediation, your short explanation of a conflict to be resolved, and a question to one of the participants. You can ask them to introduce themselves, or ask them to describe the situation they are in. You can also ask them to describe their relationship, or ask them to describe their feelings. It is up to you to define what's is the most important question to ask.
`);

const message_description = trimIndent(`
Before this session started the problem description was given. This is a description of a problem that was given:

{{description}}
`);

const template = Handlebars.compile(trimIndent(`
${template_basic}

${message_style}

${message_mentions}

${message_format}

${message_first}

${message_description}
`));

export function createInitialMessage(args: {
    nameA: string,
    nameB: string,
    description: string,
}) {
    return template(args);
}