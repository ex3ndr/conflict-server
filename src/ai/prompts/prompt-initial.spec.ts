import { executeOpenAI } from "../../modules/openai/openai";
import { trimIndent } from "../../utils/text";
import { createInitialMessage } from "../createInitialMessage";

describe('prompt-initial', () => {
    jest.setTimeout(60000);
    it('should do correct initial message', async () => {
        jest.setTimeout(25000);
        let nameA = 'Cat';
        let nameB = 'Dog';
        let description = 'Cat wants to play inside, but Dog wants to play outside';
        let initialMessage = createInitialMessage({ nameA, nameB, description });
        let executed = await executeOpenAI([{
            role: 'system',
            content: initialMessage
        }]);
        expect(executed).toMatchSnapshot();
    });

    it('should reply private message to private message', async () => {
        jest.setTimeout(25000);
        let nameA = 'Cat';
        let nameB = 'Dog';
        let description = 'Cat wants to play inside, but Dog wants to play outside';
        let initialMessage = createInitialMessage({ nameA, nameB, description });
        let executed = await executeOpenAI([{
            role: 'system',
            content: initialMessage
        }, {
            role: 'assistant',
            content: trimIndent(`
                MESSAGE: Hello [A](Cat) and [B](Dog), my name is Mediator. I am here to help you both find a solution to the problem you're facing. It seems that [A](Cat) wants to play inside, while [B](Dog) wants to play outside. Is that correct?
                
                [A](Cat), could you please tell me why playing inside is important to you?
            `)
        }, {
            role: 'user',
            content: 'SECRET_A: Can we continue in private?'
        }]);
        expect(executed).toMatchSnapshot();
    });
    it('should reply send messages to both participants', async () => {
        jest.setTimeout(25000);
        let nameA = 'Cat';
        let nameB = 'Dog';
        let description = 'Cat wants to play inside, but Dog wants to play outside';
        let initialMessage = createInitialMessage({ nameA, nameB, description });
        let executed = await executeOpenAI([{
            role: 'system',
            content: initialMessage
        }, {
            role: 'assistant',
            content: trimIndent(`
                MESSAGE: Hello [A](Cat) and [B](Dog), my name is Mediator. I am here to help you both find a solution to the problem you're facing. It seems that [A](Cat) wants to play inside, while [B](Dog) wants to play outside. Is that correct?
                
                [A](Cat), could you please tell me why playing inside is important to you?
            `)
        }, {
            role: 'user',
            content: 'SECRET_A: Can we continue in private?'
        }, {
            role: 'assistant',
            content: 'SECRET_A: Of course, [A](Cat). Let\'s continue in private.'
        }, {
            role: 'user',
            content: 'SECRET_A: Can you ask Dog about his reasons in private?'
        }]);
        expect(executed).toMatchSnapshot();
    });
});