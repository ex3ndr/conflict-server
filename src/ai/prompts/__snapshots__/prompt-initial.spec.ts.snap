// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prompt-initial should do correct initial message 1`] = `
"#PUBLIC
Hello [A](Cat) and [B](Dog). I am Mediator, and I am here to assist you both in reaching a fair and mutually beneficial solution. During this session, it's essential to be open, respectful, and direct. You both seem to have differing preferences on where to play. Let's begin by understanding your perspectives better. [A](Cat), could you please explain why you prefer to play inside? 
#SECRET_A
#SECRET_B
#END"
`;

exports[`prompt-initial should reply private message to private message 1`] = `
"#PUBLIC
#SECRET_A
Sure, [A](Cat), we can continue in private. Your comfort is important in this mediation process.
#SECRET_B
#END"
`;

exports[`prompt-initial should reply send messages to both participants 1`] = `
"#PUBLIC
#SECRET_A
Sure, [A](Cat). I will ask [B](Dog) his reasons in private.
#SECRET_B
[B](Dog), [A](Cat) wanted to know why playing outside is appealing to you. Could you please explain your reasons?
#END"
`;
